l5kit/Pipfile

```diff
@@ -0,0 +1,42 @@
+[[source]]
+name = "pypi"
+url = "https://pypi.org/simple"
+verify_ssl = true
+
+[dev-packages]
+pytest = ">=5.4.3"
+mypy = ">=0.782"
+pytest-cov = ">=2.10.0"
+flake8 = ">=3.8.3"
+isort = ">=5.0.4"
+sphinx = ">=3.1.1"
+sphinx-rtd-theme = ">=0.5.0"
+recommonmark = ">=0.6.0"
+pre-commit = ">=2.5.1"
+black = "==19.10b0"
+importlib-metadata = "*"  # required for python 3.8
+
+[packages]
+zarr = "*"
+numpy = "*"
+opencv-contrib-python-headless = "*"
+pymap3d = "*"
+scipy = "*"
+tqdm = "*"
+transforms3d = "*"
+pyyaml = "*"
+matplotlib = "*"
+imageio = "*"
+notebook = "*"
+ptable = "*"
+ipywidgets = "*"
+protobuf = ">=3.12.2"
+torch = ">=1.5.0,<1.6.0"
+torchvision = ">=0.6.0,<0.7.0"
+ipython = "<=7.16.1"
```

Comment:
This seems dangerous. Was there any incompatible change lately? If yes please add a comment

Reply:
No, but macOS has already the 7.17 while linux doesn't. If I generated the lock from MacOS without capping the version CI workers won't be able to install from the lock..

Reply:
You mean that the latest versions here https://pypi.org/project/ipython/#history are only for Mac? That seems very strange. I would assume that thing are always first available for linux and later for mac, if  this actually happens.

Reply:
I've checked now and it seems like that issue has been fixed. Let me try to remove the line and see if it passes on CI machines

Reply:
nice :)

Reply:
checked again, newer ipython are [only 3.7+](https://ipython.readthedocs.io/en/stable/whatsnew/version7.html) . I can generate the lock from a 3.6 python to ensure the locked version is 3.6-compatible, is that fine with you? The risk there is obviously someone updating the lock in the future from 3.8  or 3.7.

Alternatively, we could drop support for 3.6

Reply:
why is ipython in the pipfile? 
is there a dependency on it?

Reply:
jupyter has a dependency on ipython. If I let pipenv to manage it it will have a different behaviour depending on the python version you use to generate the lock:
- python 3.6 -> lock 7.16
- python 3.7, 3.8 -> lock 7.18 or the most recent one

The issue here is that ipython has dropped support for 3.6 between two versions, what is the best practise here?

Reply:
jupyter should resolve its dependencies on its own. II don't think you need to add the internal dependencies.
In general I don't think its a great idea to support several python versions. What is the advantage? You can just stick to one version.

Reply:
> jupyter should resolve its dependencies on its own. II don't think you need to add the internal dependencies.

It's a locking issue here. If we lock from 3.8, ipython will be locked at 7.18 which is not available on 3.6 in CI.



> In general I don't think it's a great idea to support several python versions. What is the advantage? You can just stick to one version.

Well I guess one advantage is that if you're on ubuntu 18.04LTS (default python is still 3.6.8 there) you don't have to install a new interpreter just to use our package. Apart from that, I don't see any particular issues in dropping support for 3.6 and 3.7. It will also make maintenance easier. So should we just aim to support 3.8 in the end @giacomodabisias ?